contents/blog/temporal-exports.md

```diff
@@ -0,0 +1,131 @@
+
+Although stateless plugins (like GeoIP) worked well with our [plugin service structure](/blog/how-we-built-an-app-server), **stateful** plugins (like export apps) didn’t. This is because they depended on external resources that can degrade or fail. Dealing with this required timeout and retry logic that was difficult to build into the existing structure.
+
+Export apps scaled as far as possible. They worked well for smaller users, but as the export requirements grew, it started to show issues. Chiefly among them was that they were:
+
+### 1. Unreliable
+
+The export manager had no way of reliably detecting whether an export had failed or was simply taking too long, and it would eagerly retry after a timeout.
+
+This caused multiple copies of the same export job with the same state to run at the same time. This created a race condition where whenever a new copy started, it would reset the progress to zero, and other copies would resume from the beginning again. This caused duplication in the exported user data, high infrastructure load, and database locks and conflicts.
+
+Another big issue was needing to run entire batches again if they failed. It wasn’t possible to restart from where we left off. Our team would often need to manually trigger resets. Exports couldn’t pause or run simultaneously either. 
+
+### 2. Opaque
+
+On top of unreliability, we didn’t have a clear picture of what was happening in the export apps. Throughout our codebase, we had multiple different work queues like Celery, Kafka, and Graphile each running separately in different contexts without full visibility into those contexts.
+
+We didn’t have logs or reports for export apps. For example, we didn’t know how much data was being exported or when duplicate runs were happening. When issues happened, we relied on users to tell us. This led to a lot of firefighting and often manual restarts of exports.
+
+### 3. Expensive
+
+The final nail was that export apps were expensive. This wasn’t because we were charging a lot for exports – in fact, we weren’t charging anything. It was because they were expensive on the destination side. 
+
+The unreliability led to larger amounts of data processing on the destination side, all of which come with costs. For example, if your export failed 75% through and you needed to reset, you end up paying ingestion fees of 175%. We were also paying for more processing than needed because exports were failing repeatedly.
+
+On top of extra processing time, export apps were taking up large amounts of our team’s time. Reliability needed fixing before we worked on performance, and firefighting always focused on reliability. More importantly, our team was spending large amounts of time firefighting and "babysitting" exports. They had to be reactive, rather than proactive.
```

Comment:
```suggestion
Finally, export apps were an expensive drain on the pipeline team's time. Our team was spending too much time firefighting and "babysitting" exports. They had to reactively deal with reliability problems, rather than proactively working on performance.
```